src/tls/ssl_context.rs

```diff
@@ -0,0 +1,176 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use openssl::ssl::{ Ssl, SslContext, SslMethod, SSL_VERIFY_NONE };
+use openssl::ssl::error::SslError;
+use openssl::x509::X509FileType;
+use openssl_sys;
+
+use std::collections::HashMap;
+use std::io::Error;
+use std::path::Path;
+use std::sync::{ Arc, RwLock };
+
+use tls::certificate_record::CertificateRecord;
+
+pub trait SslContextProvider : Send  {
+    fn context(&self) -> Result<SslContext, Error>;
+    fn update(&mut self, HashMap<String, CertificateRecord>) -> ();
+}
+
+#[derive(Clone)]
+pub struct SniSslContextProvider {
+    main_context: Arc<RwLock<SslContext>>
+}
+
+impl SslContextProvider for SniSslContextProvider {
+    fn context(&self) -> Result<SslContext, Error> {
+        Ok(self.main_context.read().unwrap().clone())
+    }
+
+    fn update(&mut self, configured_hosts: HashMap<String, CertificateRecord>) -> () {
+        debug!("Updating SniSslContextProvider");
+
+        let mut new_ssl_hosts = HashMap::new();
+
+        for record in configured_hosts.values() {
+            debug!("Creating SslContext for {}", record.hostname);
+            let ssl_context = create_ssl_context(&record.cert_file, &record.private_key_file);
+
+            if ssl_context.is_ok() {
+                let ssl_context = ssl_context.unwrap();
+                new_ssl_hosts.insert(record.hostname.clone(), ssl_context);
+            } else {
+                error!("Failed to configure SslContext for {} using key: {:?}, and cert: {:?}", &record.hostname, &record.cert_file, &record.private_key_file);
```

Comment:
This is a long line :)

Please, make sure that all lines are shorter than 80 chars, please.